### PR TITLE
OS-8566 want link-local route support in dockerinit

### PR DESCRIPTION
**Pull request by hww3**
Hello-

Details of the situation this aims to correct are described in #984. A previous patch for native zones was proposed and merged through issue illumos-joyent#244. This change is not cstyle clean, though it doesn't appear to be a cstyle'd source file.

Synopsis of change:

- add support for link-local routes (equivalent of using -interface flag to route add)
- order creation of routes such that link-local routes are created before any default gateway or static routes
- minor rearrangement of functions so that this behavior clear

Some testing notes:

I have a few nodes at OVH that use "failover" ip addresses. These addresses are handed out on a CIDR subnet different than the admin network. The default route, however, is that of the default network. So a link-local route needs to be created before the default route. That's the rationale for this change.

I've been using this patch on a number docker/lx zones across multiple hosts for the past year. These zones require the link local route to be created first as the default route is on the link-local network (that isn't the same as the IP address of the interface.) Without the patch, creating the default route fails with a "no route to host" error (and the zone provisioning fails after a time), but is successful with the patch.

I've also been running platform images with this fix on multiple systems that don't have static routes at all, and these all work as expected with the the patch.

Additionally, on a dummy test system, I created a simnet device and assigned a nic tag to that. I then created a series of lx/docker zones using vmadm with 2 nics each and the following combinations of routes. I used a random docker image that I already had available and used zlogin to examine the state of the routes:

- static routes without a prefix component (ie host route)
- static routes with a prefix component
- link local routes without a prefix component (ie host route)
- link local routes with a prefix component
- varying the nic that the link local route was specified over

These all seemed to work just fine, in that:

- no static routes (results in no errors)
- link-local routes in combination with a default route (no errors and link local route created first)
- ordering of non-link-local routes in combination with link-local routes (no errors and link-local route created before non-link-local route)
- all the routes were created with correct net masks
- the static routes were added as expected in the "non link-local" path.
- the link local routes were created on the correct target nice

I'm happy to provide any additional details you think would be helpful. Looking forward to your feedback!

### CODE REVIEW

**Comment by danmcd** [comment]
Cf. TritonDataCenter/illumos-joyent#244

**Review by danmcd — COMMENTED** [review]
You should add a copyright-you line to this.  If you don't want your name on it, you should put a 2024 MNX copyright in.  (Checks will barf otherwise.)

Otherwise the code seems fine to me.  I'm gonna build a branch off of it and see what happens.

**Comment by danmcd** [comment]
One more thing...

`gmake check` won't fail because of copyrights but it WILL fail because of basic cstyle hygiene.  Here's the fix:

```diff
diff --git a/src/dockerinit/src/dockerinit.c b/src/dockerinit/src/dockerinit.c
index baf3462a..0ab4075e 100644
--- a/src/dockerinit/src/dockerinit.c
+++ b/src/dockerinit/src/dockerinit.c
@@ -1136,7 +1136,8 @@ prefixToNetmask(int pfx, struct sockaddr_in *netmask_sin)
 }
 
 int
-addRoute(const char *ifname, const char *gw, const char *dst, int dstpfx, boolean_t isLinkLocal)
+addRoute(const char *ifname, const char *gw, const char *dst, int dstpfx,
+    boolean_t isLinkLocal)
 {
     int idx;
     int len;
@@ -1153,7 +1154,7 @@ addRoute(const char *ifname, const char *gw, const char *dst, int dstpfx, boolea
     (void) bzero(rtm, RTMBUFSZ);
     rtm->rtm_addrs = RTA_DST | RTA_GATEWAY | RTA_NETMASK;
 
-    if(!isLinkLocal)
+    if (!isLinkLocal)
         rtm->rtm_flags = RTF_UP | RTF_STATIC | RTF_GATEWAY;
 
     rtm->rtm_msglen = sizeof (rtbuf);
```

**Comment by hww3** [comment]
Corrected style and added copyright line, as requested.

**Review by danmcd — DISMISSED** [review]
So `gmake check` clears.  ANd the code you have indeed looks like the sibling of your LX zone one.

I need to review your test notes, as you might have already addressed my concerns about regression testing.  I *hope* @tsoome might have a look at the code, and @bahamat might have a look at the testing notes.  Test-notes review (and extra testing if need be) will enable me to IA this.

AFAICT this has a good chance of landing early next week subject to reviewer availability.  I think this has a GREAT chance of landing before next week ends, which means it'll be in for 20240822 !

**Comment by danmcd** [comment]
PI build with this fix: https://kebe.com/~danmcd/webrevs/hww3-OS-8566-platform-20240809T194106Z.tgz
MD5 == 76504e79a4010eca89d84cc24c694972

File: src/dockerinit/src/dockerinit.c [hunk 1]
```diff
@@ -756,7 +757,7 @@ setupGateway(nvlist_t *data)
             if ((ret == 0) && (primary == B_TRUE)) {
                 ret = nvlist_lookup_string(data, "gateway", &gateway);
                 if (ret == 0) {
-                    (void) addRoute(iface, gateway, "0.0.0.0", 0, B_FALSE);
+                    (void) addRoute(iface, gateway, "0.0.0.0", 0, 0);
```
[the remark this answers is not in the document]
Comment (danmcd) [reply src/dockerinit/src/dockerinit.c:760]: boolean_t already exists in this file so stdbool is a bit much for THIS change.

You very much SHOULD, however, keep B_FALSE.

[the remark this answers is not in the document]
Comment (tsoome) [reply src/dockerinit/src/dockerinit.c:760]: Yes, it makes sense to keep stdbool for followup work, I do agree.

[the remark this answers is not in the document]
Comment (hww3) [reply src/dockerinit/src/dockerinit.c]: Thanks for pointing out this omission, I've pushed a fix.

**Review by tsoome — APPROVED** [review]
Thanks!

**Comment by danmcd** [comment]
Your test notes are quite comprehensive, including your battle-testing with the two cases (normal, and with on-link routes) that matter most.

**Review by danmcd — APPROVED** [review]

**Pull request merged**